File: android/sdk/src/main/kotlin/com/pixeldive/sdk/ImageProto.kt
package com.pixeldive.sdk

import java.time.Instant
import java.util.UUID

internal object GrpcPath {
    const val UPLOAD_IMAGE = "/pixeldive.session.v1.SessionService/UploadImage"
    const val UPLOAD_BATCH = "/pixeldive.session.v1.SessionService/UploadImagesBatch"
    const val DOWNLOAD = "/pixeldive.session.v1.SessionService/DownloadImage"
}

internal data class BatchPiece(
    val sessionId: String,
    val index: Int,
    val filename: String,
    val contentType: String,
    val data: ByteArray,
    val first: Boolean,
    val end: Boolean,
)

internal object ImageProto {
    fun imageChunk(
        sessionId: String = "",
        filename: String = "",
        contentType: String = "",
        data: ByteArray,
        metadata: String = "",
    ): ByteArray {
        val parts = ArrayList<ByteArray>()
        if (sessionId.isNotEmpty()) parts.add(ProtoWire.stringField(1, sessionId))
        if (filename.isNotEmpty()) parts.add(ProtoWire.stringField(3, filename))
        if (contentType.isNotEmpty()) parts.add(ProtoWire.stringField(4, contentType))
        parts.add(ProtoWire.bytesField(5, data))
        if (metadata.isNotEmpty()) parts.add(ProtoWire.stringField(6, metadata))
        return concat(parts)
    }

    fun batchChunk(piece: BatchPiece): ByteArray {
        val parts = ArrayList<ByteArray>()
        if (piece.first && piece.sessionId.isNotEmpty()) {
            parts.add(ProtoWire.stringField(1, piece.sessionId))
        }
        if (piece.index != 0) {
            parts.add(ProtoWire.varintField(2, piece.index.toLong()))
        }
        if (piece.first && piece.filename.isNotEmpty()) {
            parts.add(ProtoWire.stringField(3, piece.filename))
        }
        if (piece.first && piece.contentType.isNotEmpty()) {
            parts.add(ProtoWire.stringField(4, piece.contentType))
        }
        parts.add(ProtoWire.bytesField(5, piece.data))
        if (piece.end) {
            parts.add(ProtoWire.boolField(6, true))
        }
        return concat(parts)
    }

    fun downloadRequest(
        sessionId: String,
        imageId: String,
    ): ByteArray = ProtoWire.stringField(1, sessionId) + ProtoWire.stringField(2, imageId)

    fun payloadChunks(
        payload: ByteArray,
        size: Int,
    ): List<ByteArray> {
        val window = size.coerceAtLeast(1)
        if (payload.isEmpty()) {
            return listOf(ByteArray(0))
        }
        val pieces = ArrayList<ByteArray>()
        var offset = 0
        while (offset < payload.size) {
            val end = (offset + window).coerceAtMost(payload.size)
            pieces.add(payload.copyOfRange(offset, end))
            offset = end
        }
        return pieces
    }

    fun decodeUpload(body: ByteArray): SessionImage =
        fieldOneImages(body).firstOrNull()
            ?: throw PixeldiveException.Decoding("gRPC upload response missing image")

    fun decodeBatch(body: ByteArray): List<SessionImage> = fieldOneImages(body)

    private fun fieldOneImages(body: ByteArray): List<SessionImage> {
        val reader = ProtoReader(body)
        val images = ArrayList<SessionImage>()
        while (true) {
            val next = reader.next() ?: break
            if (next.first == 1 && next.second == 2) {
                images.add(decodeSessionImage(next.third))
            }
        }
        return images
    }

    fun decodeChunkData(body: ByteArray): ByteArray {
        val reader = ProtoReader(body)
        var payload = ByteArray(0)
        while (true) {
            val next = reader.next() ?: break
            if (next.first == 5 && next.second == 2) {
                payload = next.third
            }
        }
        return payload
    }

    fun decodeSessionImage(body: ByteArray): SessionImage {
        val fields = ImageFields()
        val reader = ProtoReader(body)
        while (true) {
            val next = reader.next() ?: break
            fields.accept(next)
        }
        return fields.toImage()
    }
}

private class ImageFields {
    var id = ""
    var sessionId = ""
    var filename = ""
    var contentType = ""
    var sizeBytes = 0L
    var uploaded = Instant.EPOCH

    fun accept(next: Triple<Int, Int, ByteArray>) {
        val (field, wire, value) = next
        if (wire == 2) {
            acceptBytes(field, value)
            return
        }
        if (field == 5 && wire == 0) {
            sizeBytes = decodeVarint(value)
        }
    }

    private fun acceptBytes(
        field: Int,
        value: ByteArray,
    ) {
        when (field) {
            1 -> id = utf8(value)
            2 -> sessionId = utf8(value)
            3 -> filename = utf8(value)
            4 -> contentType = utf8(value)
            7 -> uploaded = decodeTimestamp(value)
        }
    }

    fun toImage(): SessionImage =
        SessionImage(
            id = UUID.fromString(ResourceId.parse(id)),
            sessionId = UUID.fromString(ResourceId.parse(sessionId)),
            filename = filename,
            contentType = contentType,
            sizeBytes = sizeBytes.toInt(),
            uploadedAt = uploaded,
            metadata = emptyMap(),
        )
}

private fun decodeTimestamp(body: ByteArray): Instant {
    val reader = ProtoReader(body)
    var seconds = 0L
    while (true) {
        val next = reader.next() ?: break
        if (next.first == 1 && next.second == 0) {
            seconds = decodeVarint(next.third)
        }
    }
    return Instant.ofEpochSecond(seconds)
}

private fun utf8(data: ByteArray): String = data.toString(Charsets.UTF_8)

private fun concat(parts: List<ByteArray>): ByteArray {
    val out = ByteArray(parts.sumOf { it.size })
    var offset = 0
    parts.forEach { part ->
        System.arraycopy(part, 0, out, offset, part.size)
        offset += part.size
    }
    return out
}

File: ios/Sources/PixeldiveSDK/ImageProto.swift
import Foundation

enum GrpcPath {
    static let uploadImage = "/pixeldive.session.v1.SessionService/UploadImage"
    static let uploadBatch = "/pixeldive.session.v1.SessionService/UploadImagesBatch"
    static let download = "/pixeldive.session.v1.SessionService/DownloadImage"
}

struct BatchPiece {
    var sessionID: String
    var index: Int
    var filename: String
    var contentType: String
    var data: Data
    var first: Bool
    var end: Bool
}

enum ImageProto {
    static func imageChunk(
        sessionID: String = "",
        filename: String = "",
        contentType: String = "",
        data: Data,
        metadata: String = ""
    ) -> Data {
        var message = Data()
        if !sessionID.isEmpty {
            message.append(ProtoWire.stringField(1, sessionID))
        }
        if !filename.isEmpty {
            message.append(ProtoWire.stringField(3, filename))
        }
        if !contentType.isEmpty {
            message.append(ProtoWire.stringField(4, contentType))
        }
        message.append(ProtoWire.bytesField(5, data))
        if !metadata.isEmpty {
            message.append(ProtoWire.stringField(6, metadata))
        }
        return message
    }

    static func batchChunk(_ piece: BatchPiece) -> Data {
        var message = Data()
        if piece.first, !piece.sessionID.isEmpty {
            message.append(ProtoWire.stringField(1, piece.sessionID))
        }
        if piece.index != 0 {
            message.append(ProtoWire.varintField(2, UInt64(piece.index)))
        }
        if piece.first, !piece.filename.isEmpty {
            message.append(ProtoWire.stringField(3, piece.filename))
        }
        if piece.first, !piece.contentType.isEmpty {
            message.append(ProtoWire.stringField(4, piece.contentType))
        }
        message.append(ProtoWire.bytesField(5, piece.data))
        if piece.end {
            message.append(ProtoWire.boolField(6, true))
        }
        return message
    }

    static func downloadRequest(sessionID: String, imageID: String) -> Data {
        var message = Data()
        message.append(ProtoWire.stringField(1, sessionID))
        message.append(ProtoWire.stringField(2, imageID))
        return message
    }

    static func payloadChunks(_ payload: Data, size: Int) -> [Data] {
        let window = max(size, 1)
        if payload.isEmpty {
            return [Data()]
        }
        var pieces: [Data] = []
        var offset = 0
        while offset < payload.count {
            let end = min(offset + window, payload.count)
            pieces.append(payload.subdata(in: offset..<end))
            offset = end
        }
        return pieces
    }

    static func decodeUpload(_ body: Data) throws -> SessionImage {
        var reader = ProtoReader(bytes: body)
        while let next = try reader.next() {
            if next.number == 1, next.wire == 2 {
                return try decodeSessionImage(next.value)
            }
        }
        throw PixeldiveError.decoding("gRPC upload response missing image")
    }

    static func decodeBatch(_ body: Data) throws -> [SessionImage] {
        var reader = ProtoReader(bytes: body)
        var images: [SessionImage] = []
        while let next = try reader.next() {
            if next.number == 1, next.wire == 2 {
                images.append(try decodeSessionImage(next.value))
            }
        }
        return images
    }

    static func decodeChunkData(_ body: Data) throws -> Data {
        var reader = ProtoReader(bytes: body)
        var payload = Data()
        while let next = try reader.next() {
            if next.number == 5, next.wire == 2 {
                payload = next.value
            }
        }
        return payload
    }

    static func decodeSessionImage(_ body: Data) throws -> SessionImage {
        var reader = ProtoReader(bytes: body)
        var id = ""
        var sessionID = ""
        var filename = ""
        var contentType = ""
        var sizeBytes: Int64 = 0
        var uploaded = Date(timeIntervalSince1970: 0)
        while let next = try reader.next() {
            switch (next.number, next.wire) {
            case (1, 2):
                id = string(next.value)
            case (2, 2):
                sessionID = string(next.value)
            case (3, 2):
                filename = string(next.value)
            case (4, 2):
                contentType = string(next.value)
            case (5, 0):
                sizeBytes = Int64(try decodeVarint(next.value))
            case (7, 2):
                uploaded = try decodeTimestamp(next.value)
            default:
                continue
            }
        }
        return SessionImage(
            id: try parseUUID(id),
            sessionId: try parseUUID(sessionID),
            filename: filename,
            contentType: contentType,
            sizeBytes: Int(sizeBytes),
            uploadedAt: uploaded,
            metadata: [:]
        )
    }
}

private func string(_ data: Data) -> String {
    String(data: data, encoding: .utf8) ?? ""
}

private func parseUUID(_ raw: String) throws -> UUID {
    let parsed = try ResourceID.parse(raw)
    guard let uuid = UUID(uuidString: parsed) else {
        throw PixeldiveError.invalidResourceID(raw)
    }
    return uuid
}

private func decodeTimestamp(_ body: Data) throws -> Date {
    var reader = ProtoReader(bytes: body)
    var seconds: Int64 = 0
    while let next = try reader.next() {
        if next.number == 1, next.wire == 0 {
            seconds = Int64(try decodeVarint(next.value))
        }
    }
    return Date(timeIntervalSince1970: TimeInterval(seconds))
}
